### PR TITLE
Add Platform Reliability RFC

### PR DESCRIPTION
**Pull request by kenoir**
Following our discussion, i've aggregated what we talked about into some more general proposals that we can flesh out and work on. 

I've just tried to summarise what we said, not create specific issues to be worked on. I want to use these PR to make sure we agree in principle on the general direction, and to make sure nothing important has been missed.

When we merge this PR I will create issues from the higher level actions and add them to:

https://github.com/wellcometrust/platform/projects/12 

Uncategorised stuff has already had issues created in that project:
- https://github.com/wellcometrust/platform/issues/4237
- https://github.com/wellcometrust/platform/issues/4236
- https://github.com/wellcometrust/platform/issues/4235
- https://github.com/wellcometrust/platform/issues/4234

### CODE REVIEW

File: rfcs/019-platform_reliability/observability.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+- Implement a tracing solution in our piplelines that allow us to visualise the flow of work through our services.
+
+## Metrics
+
+In order that application issues are quickly found and diagnosed we should be able to view metrics from our applications easily.
+
+We should:
+
+- Decide on appropriate metric collectors for each of our products.
+- Decide on a single platform for visualising application metrics across our products.
```
Comment (davidpmccormick) [inline rfcs/019-platform_reliability/observability.md]: Is a single platform necessary/realistic? Or could it be a single place from which to access all metric visualising things (possibly residing on different platforms)?

**Comment by jamesgorrie** [comment]
We have a project in the experience repo:
https://github.com/wellcometrust/wellcomecollection.org/projects/42

I prefer the issues to live as close to their code as possible. We can have project boards with other repo's issues on them, I just haven't found the owner of `platform` to do so.

@pollecuttn and I tried, no luck - maybe @jtweed is the owner?

**Comment by jtweed** [comment]
Not quite following what you mean, but in any case I think I'm an admin on platform but not the owner.

File: rfcs/019-platform_reliability/ci_cd.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+## Deployment
+
+###  Continuous deployment
+
+When test pass against a build in a CI environment those changes are deployed to production.
+
+We need:
+- To be satisfied that tests in CI ensure that a build is promotable to production.
+- To build on the release tool to ensure the deploy step can actually trigger deployments.
+- To connect tests to deployment!
```
Comment (jamesgorrie) [inline rfcs/019-platform_reliability/ci_cd.md]: Can we add something here about rolling back, as if we have CD, the confidence in reverting quick is necessary.

File: rfcs/019-platform_reliability/ci_cd.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
```
Comment (jamesgorrie) [inline rfcs/019-platform_reliability/ci_cd.md]:
```suggestion
- To be able to roll back to a know working version quickly (quickly to be defined)
```

File: rfcs/019-platform_reliability/ci_cd.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+We want to be able to see who deployed what & when in the recent past.
+
+We should provide an authenticated web dashboard showing deployments
+We want to know:
+- What was deployed, referencing:
+  - The PR that was deployed
+  - The tests that have been run against the deployed changes (pre/post deployment)
+  - The status of a deployment
+- When was a change deployed
+
```
Comment (jamesgorrie) [inline rfcs/019-platform_reliability/ci_cd.md]:
```suggestion
- When a change is available but not deployed
```

File: rfcs/019-platform_reliability/ci_cd.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+We need:
+- To be satisfied that tests in CI ensure that a build is promotable to production.
+- To build on the release tool to ensure the deploy step can actually trigger deployments.
+- To connect tests to deployment!
+
+### Deployment visibility
+
+We want to be able to see who deployed what & when in the recent past.
+
+We should provide an authenticated web dashboard showing deployments
```
Comment (jamesgorrie) [inline rfcs/019-platform_reliability/ci_cd.md]:
```suggestion
We should provide an authenticated web dashboard showing deployments.

```

File: rfcs/019-platform_reliability/ci_cd.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+When test pass against a build in a CI environment those changes are deployed to production.
+
+We need:
+- To be satisfied that tests in CI ensure that a build is promotable to production.
+- To build on the release tool to ensure the deploy step can actually trigger deployments.
+- To connect tests to deployment!
+
+### Deployment visibility
+
+We want to be able to see who deployed what & when in the recent past.
```
Comment (jamesgorrie) [inline rfcs/019-platform_reliability/ci_cd.md]:
```suggestion
We want to be able to see who recently deployed what & when.
```

File: rfcs/019-platform_reliability/ci_cd.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  - The PR that was deployed
+  - The tests that have been run against the deployed changes (pre/post deployment)
+  - The status of a deployment
+- When was a change deployed
+
+### Deployment tool 
+
+The release tooling as currently used is too complicated.
+
+We should have a single step deployment tool that takes a deploy-able artifact from a build and deploys it to an environment.
```
Comment (jamesgorrie) [inline rfcs/019-platform_reliability/ci_cd.md]:
```suggestion
We should have a single step deployment tool that takes deploy-able artifact(s) from a build and deploys it to an environment.
```

Updated this, but not sure if it's clear as we need to deploy stacks, which are multiple services, and artifacts.

File: rfcs/019-platform_reliability/ci_cd.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+```
+> cd my_project
+> wellcome_release_tool deploy build123 staging
+Congratulations abc123 (PR123, PR345) deployed to staging!
+>
```
Comment (jamesgorrie) [inline rfcs/019-platform_reliability/ci_cd.md]: ⛵️

File: rfcs/019-platform_reliability/ci_cd.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+> wellcome_release_tool deploy build123 staging
+Congratulations abc123 (PR123, PR345) deployed to staging!
+> 
+```
+
+## Continuous integration
+
+In order to ensure that developers are able to work quickly and effectively together we should:
+
+- Reduce build times
```
Comment (jamesgorrie) [inline rfcs/019-platform_reliability/ci_cd.md]: I've seen it before that tests won't pass if you pass a budget, not sure if this is too dogmatic, but interesting.

File: rfcs/019-platform_reliability/observability.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+## Alerting
+
+In order that we can quickly react to application issues we should be notified when issues requiring our attention arise.
+
+We should:
+
+- Be able to track response to alerts and actions taken to resolve them.
```
Comment (jamesgorrie) [inline rfcs/019-platform_reliability/observability.md]: Not sure what this sentence means?

File: rfcs/019-platform_reliability/reliability.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+## Pipeline reliability
+
+In order to verify that pipeline deployments will behave and are behaving as expected we should:
+
+- Test pipeline changes that require re-indexes before new indexes are promoted into production view.
+
+- Test the pipeline is performing as expected when updates flow through it both:
+  - In production
+  - In integration tests
+
```
Comment (jamesgorrie) [inline rfcs/019-platform_reliability/reliability.md]: This feels like a reactionary bit of documentation that doesn't have any action bar "make sure you do it properly". 

But the integration tests between services that make up a stack feel important.

**Pull request merged**